### ISSUE TITLE
[switch-exhaustiveness-check] Require default case if discriminant type is not a union

### ISSUE THREAD

**Issue by raymondwang** [typescript-eslint/typescript-eslint#2959]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": "warn"
  }
}
```

```TS
type Day = string; // or number

const day = 'Monday' as Day;
let result = 0;

switch (day) {
  case 'Monday': {
    result = 1;
    break;
  }
}
```
or
```TS
type Day = any; // or unknown, or a JS variable outside of a TS context

const day = 'Monday' as Day;
let result = 0;

switch (day) {
  case 'Monday': {
    result = 1;
    break;
  }
}
```

**Expected Result**

I would expect to see an error message compatible with the [`default-case` ESLint rule](https://eslint.org/docs/rules/default-case), indicating that the switch-exhaustiveness-check rule is invalidated because the expression is not a union type.

**Actual Result**

No error message.

**Additional Info**

Essentially, I think the best of both worlds is to just have the predictable functionality of the `default-case` rule operate normally except in the event that a switch statement's expression is a union type. Today, there doesn't seem to be any way for the two rules to operate in tandem. Adding this functionality might be a breaking change, but I'm honestly a bit curious how controversial it would be.

**Comment by bradzacher** [comment]
As it stands right now - this rule only checks union types.
So this is a large expansion in responsibilities for the rule.

<opinion>
I'll be honest when I say that I disagree that `default-case` is a "good" rule to have on.
There are cases where you know the constrained value set, but can't constrain the type (eg external library types), meaning you're working with a generic `string` or `number`.

So adding a `default` there just means you're going to add `default: throw new Error("shouldn't happen");`, which isn't a *great* practice by default.

Adding a `default` case also means that if you are able to convert the type to a union type in future, then you'll never catch any missing cases, as the rule would detect that all cases are handled.
</opinion>

I'm not sure if this is a good addition to this lint rule.
*Maybe* behind a default false flag? But that's, at best, what I'd be suggesting.

**Comment by raymondwang** [comment]
Thanks for the response @bradzacher. I think that's totally fair, and I don't want to overcomplicate this process or rule by enforcing my opinions with a breaking change. I basically just want to implement some solution for codebases that _do_ want to enforce `default-case` as well as `switch-exhaustiveness-check`, which are totally incompatible rules today (despite both being in the interest of encouraging switch statement thoroughness).

Is there a way to recognize whether the `default-case` rule is being enforced within the rule? Could be an elegant solution, rather than passing in another config, but obviously happy to do the latter if now.

**Comment by bradzacher** [comment]
> Is there a way to recognize whether the default-case rule is being enforced within the rule?

No. Lint rules are all isolated. They have no way to tell if other rules are enabled.

---

Happy to have this behind a default off option!

**Comment by SimonSimCity** [comment]
Just a work-around I've found which works for me:

If you are the type of guy who wants this feature, it might be worth thinking about an exhaustive matching guard (https://medium.com/technogise/type-safe-and-exhaustive-switch-statements-aka-pattern-matching-in-typescript-e3febd433a7a#be77). This way, you keep `default-case` active and use the fact that TypeScript has this special type `never`. In a case where you have an exhaustive switch-case, let the compiler ensure the type is `never`. Here's an example:

```ts
enum Direction {
  Up,
  Down,
  Left,
  Right,
}

(myVar: Direction) => {
  switch (myVar) {
    case Direction.Up:
      break;
    case Direction.Down:
      break;
    case Direction.Left:
      break;
    case Direction.Right:
      break;
    default:
      ((_: never) => { })(myVar)
  }
}
```

[Playground](https://typescript-eslint.io/play/#ts=4.5.5&jsx=true&sourceType=module&code=KYOwrgtgBAIglgJ2AYwC5wPYigbwFBRQCqADgDQGwYDuIFhAMsAGar1QBKcA5gBZt4AvnjwAKCAE8AagEMEALliIU6LAEooAXgB8uSgGdqcVMl5Rx0uRvyFCyGfuBKkaTCAB0peZVtQARkgyANYA3D5Q9o7OKm7uMDQg3r6EAcDBYb6RTvAuqh5MrEnJqenhWdGuWO5cfKhFviWh4QAmLDJgADZ14YSiogD6iiDAAG7ACBo6uFCCahayE5TCgkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQCaIDMBDBAFwFoBjI5RdKRaaAe2knDAF8ROg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by ST-DDT** [comment]
I'll have a look on this.

**Comment by ST-DDT** [comment]
PR created: https://github.com/typescript-eslint/typescript-eslint/pull/7880

- https://github.com/typescript-eslint/typescript-eslint/pull/7880
